src/interface/state_descriptor.cpp

```diff
@@ -0,0 +1,379 @@
+    auto dependency = metadata.Role();
+    bool is_sparse = metadata.IsSet(Metadata::Sparse);
+    int sparse_id = metadata.GetSparseId();
+    if (dependency == Metadata::Private) {
+      pvp->AddPrivate(package, var, metadata);
+    } else if (dependency == Metadata::Provides) {
+      if (Provided(var)) {
+        if (is_sparse) { // will simplify with dense on block
+          if (sparse_added_provides[var][sparse_id]) {
+            std::stringstream ss;
+            ss << "Sparse variable " << var << "_" << sparse_id
+               << " provided by multiple packages" << std::endl;
+            PARTHENON_THROW(ss);
+          }
+        } else {
+          PARTHENON_THROW("Variable " + var + " provided by multiple packages");
+        }
+      }
+      if (is_sparse) sparse_added_provides[var][sparse_id] = true;
+      provided_vars.insert(var);
+      pvp->AddProvides(package, var, metadata);
+    } else if (dependency == Metadata::Requires) {
+      depends_vars.insert(var);
+    } else if (dependency == Metadata::Overridable) {
+      if (overridable_meta.count(var) == 0) {
+        overridable_meta[var] = {metadata};
+        if (is_sparse) sparse_added_depends[var][sparse_id] = true;
+      }
+      if (is_sparse) {
+        if (!sparse_added_depends[var][sparse_id]) {
+          overridable_meta[var].push_back(metadata);
+          sparse_added_depends[var][sparse_id] = true;
+        }
+      }
+      // only update overridable_vars count once
+      if (overridable_meta.at(var).size() == 1) {
+        overridable_vars[var] += 1; // using value initalization of ints = 0
+      }
+    } else {
+      PARTHENON_THROW("Unknown dependency");
+    }
+  }
+
+  template <typename Collection>
+  void CategorizeCollection(const std::string &package, const Collection &c,
+                            VariableProvider *pvp) {
+    for (auto &pair : c) {
+      std::string const &var = pair.first;
+      auto &metadata = pair.second;
+      Categorize(package, var, metadata, pvp);
+    }
+  }
+
+  void CheckRequires() {
+    for (auto &v : depends_vars) {
+      if (!Provided(v) && overridable_vars.count(v) == 0) {
+        std::stringstream ss;
+        ss << "Variable " << v
+           << " registered as required, but not provided by any package!" << std::endl;
+        PARTHENON_THROW(ss);
+      }
+    }
+  }
+
+  void CheckOverridable(VariableProvider *pvp) {
+    std::unordered_set<std::string> cache;
+    for (auto &pair : overridable_vars) {
+      auto &var = pair.first;
+      auto &count = pair.second;
+      if (!Provided(var)) {
+        if (count > 1) {
+          std::stringstream ss;
+          ss << "Variable " << var
+             << " registered as overridable multiple times, but never provided."
+             << " This results in undefined behaviour as to which package will provide"
+             << " it." << std::endl;
+          PARTHENON_DEBUG_WARN(ss);
```

Comment:
This is intentional. My thinking is that, for example, three packages might all want, say, Pressure but all register it as overridable. It feels to me like that should run, but you need to pick one to provide. Only one of them should provide it and there's no way to automatically pick. That's why the warning.

We could change it to an error, that would certainly enforce being more deliberate. But it would make it harder to "flawlessly" turn off some package which provides.

Reply:
I didn't consider it from that angle yet.
I guess my main concern is related to #396 in the sense that, if everyone overrides, multiple different definitions may float around and it's not guaranteed which one is picked.

Reply:
Ah, yes. This is an issue. How about this? I'll make a comment on the same issue, and submit a PR later to make sure dimensions, flags, etc, are consistent.

Reply:
Sounds good.